proxy/gateway/native/src/proxy_gateway.c

```diff
@@ -0,0 +1,722 @@
+        msg_size = Message_ToByteArray(message, NULL, 0);
+        if (msg_size < 0)
+        {
+            /*Codes_SRS_BROKER_13_053: [This function shall return BROKER_ERROR if an underlying API call to the platform causes an error or BROKER_OK otherwise.]*/
+            LogError("unable to serialize a message [%p]", msg);
+            Message_Destroy(msg);
+            result = BROKER_ERROR;
+        }
+        else
+        {
+            /*Codes_SRS_BROKER_17_025: [ Broker_Publish shall allocate a nanomsg buffer the size of the serialized message + sizeof(MODULE_HANDLE). ]*/
+            buf_size = msg_size;
+            void* nn_msg = nn_allocmsg(buf_size, 0);
+            if (nn_msg == NULL)
+            {
+                /*Codes_SRS_BROKER_13_053: [This function shall return BROKER_ERROR if an underlying API call to the platform causes an error or BROKER_OK otherwise.]*/
+                LogError("unable to serialize a message [%p]", msg);
+                result = BROKER_ERROR;
+            }
+            else
+            {
+                unsigned char *nn_msg_bytes = (unsigned char *)nn_msg;
+                /*Codes_SRS_BROKER_17_027: [ Broker_Publish shall serialize the message into the remainder of the nanomsg buffer. ]*/
+                Message_ToByteArray(message, nn_msg_bytes, msg_size);
+
+                /*Codes_SRS_BROKER_17_010: [ Broker_Publish shall send a message on the publish_socket. ]*/
+                int nbytes = nn_send(remote_module->message_socket, &nn_msg, NN_MSG, 0);
+                if (nbytes != buf_size)
+                {
+                    /*Codes_SRS_BROKER_13_053: [This function shall return BROKER_ERROR if an underlying API call to the platform causes an error or BROKER_OK otherwise.]*/
+                    LogError("unable to send a message [%p]", msg);
+                    /*Codes_SRS_BROKER_17_012: [ Broker_Publish shall free the message. ]*/
+                    nn_freemsg(nn_msg);
+                    result = BROKER_ERROR;
+                }
+                else
+                {
+                    result = BROKER_OK;
+                }
+            }
+            /*Codes_SRS_BROKER_17_012: [ Broker_Publish shall free the message. ]*/
+            Message_Destroy(msg);
+            /*Codes_SRS_BROKER_17_011: [ Broker_Publish shall free the serialized message data. ]*/
+        }
+
+    }
+
+    /*Codes_SRS_BROKER_13_037: [ This function shall return BROKER_ERROR if an underlying API call to the platform causes an error or BROKER_OK otherwise. ]*/
+    return result;
+}
+
+
+int
+connect_to_message_channel (
+    REMOTE_MODULE_HANDLE remote_module,
+    const MESSAGE_URI * channel_uri
+) {
+    int result;
+
+    /* SRS_PROXY_GATEWAY_027_0xx: [`connect_to_message_channel` shall create a socket for the Azure IoT Gateway message channel by calling `int nn_socket(int domain, int protocol)` with `AF_SP` as `domain` and `MESSAGE_URI::uri_type` as `protocol`] */
```

Comment:
These aren't publicly exposed. They are the details of nested function implementation. I have them ready for promotion if need be, but they are unnumbered until they are promoted.
Currently, they are referred to in bulk as public requirements. I was trying something new, we should address the final handling as a group, and I have them ready to copy/paste once a decision is made.